File: backend/src/api/wire/guild.rs
use serde::{Deserialize, Serialize};
use twilight_model::channel::message::sticker::StickerFormatType;
use twilight_model::channel::ChannelType;

use twilight_model::id::marker::{
    ChannelMarker, EmojiMarker, GuildMarker, RoleMarker, StickerMarker,
};
use twilight_model::id::Id;

use crate::bot::cache::{CacheChannel, CacheEmoji, CacheGuild, CacheRole};

#[derive(Debug, Serialize, Deserialize)]
pub struct GuildWire {
    pub id: Id<GuildMarker>,
    pub name: String,
    pub icon: Option<String>,
}

impl From<&CacheGuild> for GuildWire {
    fn from(g: &CacheGuild) -> Self {
        Self {
            id: g.id,
            name: g.name.clone(),
            icon: g.icon.as_ref().map(|h| h.to_string()),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GuildRoleWire {
    pub id: Id<RoleMarker>,
    pub name: String,
    pub managed: bool,
}

impl From<&CacheRole> for GuildRoleWire {
    fn from(r: &CacheRole) -> Self {
        Self {
            id: r.id,
            name: r.name.to_string(),
            managed: r.managed,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GuildChannelWire {
    pub id: Id<ChannelMarker>,
    pub name: Option<String>,
    pub position: Option<i32>,
    pub parent_id: Option<String>,
    #[serde(rename = "type")]
    pub kind: ChannelType,
}

impl From<&CacheChannel> for GuildChannelWire {
    fn from(c: &CacheChannel) -> Self {
        Self {
            id: c.id,
            name: c.name.clone(),
            position: c.position,
            parent_id: c.parent_id.map(|p| p.to_string()),
            kind: c.kind,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GuildEmojiWire {
    pub id: Id<EmojiMarker>,
    pub name: String,
    pub available: bool,
    pub animated: bool,
    pub managed: bool,
}

impl From<&CacheEmoji> for GuildEmojiWire {
    fn from(e: &CacheEmoji) -> Self {
        Self {
            id: e.id,
            name: e.name.clone(),
            available: e.available,
            animated: e.animated,
            managed: e.managed,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GuildStickerWire {
    pub id: Id<StickerMarker>,
    pub name: String,
    pub description: String,
    pub available: bool,
    pub format_type: StickerFormatType,
}

File: embedg-server/api/wire/guild.go
package wire

import (
	"gopkg.in/guregu/null.v4"
)

type GuildWire struct {
	ID   string      `json:"id"`
	Name string      `json:"name"`
	Icon null.String `json:"icon"`
}

type ListGuildsResponseWire APIResponse[[]GuildWire]

type GetGuildResponseWire APIResponse[GuildWire]

type GuildChannelWire struct {
	ID       string      `json:"id"`
	Name     string      `json:"name"`
	Position int         `json:"position"`
	ParentID null.String `json:"parent_id"`
	Type     int         `json:"type"`
}

type ListChannelsResponseWire APIResponse[[]GuildChannelWire]

type GuildRoleWire struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Managed  bool   `json:"managed"`
	Color    int    `json:"color"`
	Position int    `json:"position"`
}

type ListRolesResponseWire APIResponse[[]GuildRoleWire]

type GuildEmojiWire struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Available bool   `json:"available"`
	Animated  bool   `json:"animated"`
	Managed   bool   `json:"managed"`
}

type ListEmojisResponseWire APIResponse[[]GuildEmojiWire]

type GuildStickerWire struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Available   bool   `json:"available"`
	FormantType int    `json:"formant_type"`
}

type ListStickersResponseWire APIResponse[[]GuildStickerWire]

type GuildBrandingWire struct {
	DefaultUsername  null.String `json:"default_username"`
	DefaultAvatarURL null.String `json:"default_avatar_url"`
}

type GetGuildBrandingResponseWire APIResponse[GuildBrandingWire]
